### ISSUE TITLE
Save and restore the R5900 specific register extensions

### ISSUE THREAD

**Issue by frno7** [frno7/linux#5]
The R5900 extends the MIPS general purpose registers (GPRs) from 64 bits to 128 bits. The upper 64 bits of the GPRs are used by the R5900 specific quad load/store (LQ/SQ) and multimedia instructions (MMIs).

The special HI1 and LO1 registers, which are the upper 64 bits of each of the 128-bit HI and LO registers, are used by the R5900 specific multiply and divide instructions, such as MULT1, MULTU1, DIV1, DIVU1, MADD1 and MADDU1, as well as MFHI1, MFLO1, MTHI1 and MTLO1.

Finally, a R5900 specific shift amount (SA) register contains the shift amount for the 256-bit funnel shift instruction QFSRV.

To use the instructions mentioned above in user space, these R5900 specific registers must be saved and restored appropriately by the kernel when switching contexts. If the registers are not saved and restored, they ought to be cleared to avoid leaking information between processes.

Will these registers be saved and restored for both [the o32 and the n32 Linux ABIs](https://www.linux-mips.org/wiki/MIPS_ABI_History)? It would be most useful and least surprising to support both ABIs. Of particular note is that the MMIs cannot be disabled by the kernel. The MMIs would therefore appear to mostly work, even when the kernel actually does not save and restore the registers. Problems would manifest as random register corruptions, that can be very difficult and frustrating to find.

See appendix B in the [TX79 manual](https://wiki.qemu.org/File:C790.pdf).

**Comment by sp193** [comment]
If I remember correctly, the R5900 support which Jurgen (Mega Man) committed uses n32 ABI.
As of today, support for the R5900 within GCC is incomplete. Support for the FPU did not work as of 2015 (configuration problem within libgcc) and GCC does not support any of the non-standard functions (2nd integer pipeline & MMI).

As for dealing with the custom 128-bit registers & MMI: the recommendation I got from the GCC folks and agreed with, was to use the 128-bit MMI as a hardware vector mode implementation. A number of instructions fit nicely with GCC vector support and GCC would recognize that hardware registers can only be used for either integer arithmetic or vector computations. It makes sense anyway, since there are no MMI for doing some basic things like loading 128-bit integer literals, shifting, 256-bit multiplication or division, which GCC requires for treating the GPRs as truly 128-bit.

However, I could not put together a working system.

As for the 2nd integer pipeline: there may be a need to implement some scheduler to determine which integer pipeline to use. I'm not sure why, but the homebrew GCC v3.2.2 port did not have such a mechanism and putting the predicates for the two HILO pairs seems to cause GCC to automatically choose one. It seems to be random, but I haven't seen it make conflicting choices yet. Since there are two pipelines, there is also a need to make GCC track the two pipeline hazards individually.
I've got no confidence that I did it right, so GCC still doesn't support the two integer pipelines.

If you don't intend to support any of these, I think it should still work and there should be no security risk either (you generally need to copy values to the upper 64-bits, for there to be some value there)... but you will not be using the full power of the R5900.

**Comment by frno7** [comment]
I think assembly code and C intrinsics are likely uses of MMIs and other R5900-specific extensions with Linux. GCC auto-vectorisation would obviously be very nice to have too. GCC does actually generate the R5900-specific three-operand instructions MULT and MULTU, which is why [R5900 QEMU](https://github.com/frno7/qemu) supports it in commit [21e8e8b230](https://github.com/frno7/qemu/commit/21e8e8b230af38b6bd8c953fa5f31e4a5a128e1c). See #3 for a separate issue about the R5900 FPU.

**Comment by mirh** [comment]
@uyjulian @ZirconiumX ¿

**Comment by uyjulian** [comment]
I've never really worked with PS2 Linux ABI before.  
However, it would be nice to have a consistent documented implementation for saving/restoring MMI registers.

**Comment by frno7** [comment]
@uyjulian, the 32-bit o32 ABI uses SW/LW instructions to save and restore general purpose registers (GPRs) between context switches. The 64-bit n32 ABI uses SD/LD. For the R5900 Linux kernel, I think we should always use SQ/LQ to handle the full 128-bit multimedia registers (MMRs), which otherwise would be lost. The R5900 specific shift amount (SA) register needs to be handled by the kernel as well.

If done properly, Linux applications need not worry about these details, because the MMRs would retain their values after context switches.

**Comment by Ravenslofty** [comment]
We could get into all sorts of ABI disputes, but I think N32 is strictly
superior to O32.

The FPU state for context switch should also be preserved, including the
FPU accumulator, which has no direct instructions for accessing it.

**Comment by frno7** [comment]
> We could get into all sorts of ABI disputes, but I think N32 is strictly superior to O32.

MIPS Linux kernel maintainers have suggested that the o32 ABI must be merged before proceeding to the n32 ABI, because the latter is substantially more problematic for the R5900. I hope we can support both ABIs, so that they can be chosen freely.

I fixed the R5900 o32 ABI in Glibc 2.29. Older Glibc versions can be used if commit [8e3c00db16fc](https://sourceware.org/git/?p=glibc.git;a=commit;h=8e3c00db16fc) (MIPS: Use `.set mips2' to emulate LL/SC for the R5900 too) is applied. Glibc needs another patch to support the R5900 n32 ABI as well, but I have not submitted it yet. Both ABIs can the be emulated with an appropriately patched [R5900 QEMU](https://github.com/frno7/qemu), which is useful to compile for example a modern R5900 Gentoo Linux.

> The FPU state for context switch should also be preserved, including the FPU accumulator, which has no direct instructions for accessing it.

Do you have links to documents describing the details of this? (There are further notes about the R5900 FPU in #3.)

**Comment by Ravenslofty** [comment]
On Sat, 13 Apr 2019, 19:35 frno7, <notifications@github.com> wrote:

> We could get into all sorts of ABI disputes, but I think N32 is strictly
> superior to O32.
>
> MIPS Linux kernel maintainers have suggested that the o32 ABI must be
> merged before proceeding to the n32 ABI, because the latter is
> substantially more problematic for the R5900. I hope we can support both
> ABIs, so that they can be chosen freely.
>

That sounds reasonable, although we'll need a toolchain before we get
anywhere serious with it.

I fixed the R5900 o32 ABI in Glibc 2.29. Older Glibc versions can be used
> if commit 8e3c00db16fc
> <https://sourceware.org/git/?p=glibc.git;a=commit;h=8e3c00db16fc> (MIPS:
> Use `.set mips2' to emulate LL/SC for the R5900 too) is applied. Glibc
> needs another patch to support the R5900 n32 ABI as well, but I have not
> submitted it yet. Both ABIs can the be emulated with an appropriately
> patched R5900 QEMU <https://github.com/frno7/qemu>, which is useful to
> compile for example a modern R5900 Gentoo Linux.
>

I suspect we might not get much upstream support from them though.

The FPU state for context switch should also be preserved, including the
> FPU accumulator, which has no direct instructions for accessing it.
>
> Do you have links to documents describing the details of this? (There are
> further notes about the R5900 FPU in #3
> <https://github.com/frno7/linux/issues/3>.)
>

EE Core User's Manual, Chapter 8.

>

**Comment by frno7** [comment]
> That sounds reasonable, although we'll need a toolchain before we get anywhere serious with it.

Modern GAS and GCC are in shape and usable for R5900 o32 and n32, as well as modern R5900 Linux 5.x kernels, and for R5900 Linux distributions such as Gentoo. Toolchain patches were merged upstream last year.

In addition, a new `-mfix-r5900` GAS and GCC option is now available to compile generic MIPS Linux programs and libraries that also work with the R5900, which otherwise is incompatible due to the short loop hardware bug described in #8.

> I suspect we might not get much upstream support from them though.

A Glibc patch for the R5900 n32 ABI is forthcoming. In the meantime I have a workaround patch. I have been busy with the [R5900 QEMU](https://github.com/frno7/qemu) and the R5900 Linux kernel. I have discussed the R5900 n32 ABI with Glibc developers, and I actually believe that they will merge an appropriate patch. It is a bit more work compared with the already supported R5900 o32 ABI, though.

**Comment by sp193** [comment]
FPU accumulator, as in $ACC? There exists the sa instruction - Save Accumulator. It exists solely for the preservation of thread contexts.

EDIT: Oh, I think I remember what that was. Maybe you could use an instruction like adda, to get the current value of the FPU accumulator.

**Comment by frno7** [comment]
Review

- https://github.com/frno7/linux/discussions/90#discussioncomment-15618737
- https://github.com/frno7/linux/discussions/90#discussioncomment-15620159
- https://github.com/frno7/linux/discussions/90#discussioncomment-15622084

for implementation details!

**Comment by frno7** [comment]
> But it could be easily made with `sed` or suchlike, no fancy tools needed, and would be easy to verify, by making sure no change has been made to the machine code produced.

@maciej-w-rozycki, hmm. I’m not sure it’s easy to verify though?

There are [kernel notes on reproducible builds](https://docs.kernel.org/kbuild/reproducible-builds.html), of course, but in this case changes in line numbering in header files also seem to matter (presumably due to some debugging facilities built into the kernel). So I’m squeezing in the `gprs` macro in this manner (without surrounding whitespace) to not change the line numbering in `arch/mips/include/asm/ptrace.h`:

```c
	unsigned long __last[0];
} __aligned(8);
#define gprs(i) regs[i]
static inline unsigned long kernel_stack_pointer(struct pt_regs *regs)
{
	return regs->gprs(31);
```

Some build differences vanish by that, but some other curious cases remain. Are you building without debug or other special flags for verification purposes?

I’ve been comparing both individual `.o` files and `vmlinux` as a whole (where I would expect minor differences in timestamps etc. as per the kernel notes on reproducible builds mentioned).

**Comment by frno7** [comment]
@maciej-w-rozycki, commit b8c4c41b9368641409ec99f6c483479bbc3e3a5d is a trivial subset of the whole thing. `arch/mips/mm/fault.o` rebuilds identically, but `arch/mips/kernel/process.o` doesn’t, curiously. As mentioned, if line numbers in `arch/mips/include/asm/ptrace.h` are changed, there will be additional object file differences too.

**Comment by frno7** [comment]
@maciej-w-rozycki, I’m a quarter through by now, I think.

1. Commit be4c4766b435ba36b79021bec2fdc281a93ccf21 introduces the `gprs` macro. By far the largest patch. Mechanical replacements, as you suggested. System boots and seems fine. (It’s supposed to compile identically too.)
2. Commit 431277c1e1591b9fd88587e6ee8906f71784f7b5 introduces the 128-bit multimedia registers, without actually using them. Boots and works too. `PT_Rxx` now increments by 16 bytes (rather than 4 bytes) in `include/generated/asm-offsets.h`, as expected. So the register array enlargement and change of register offsets seems to work as well. I kept the `regs` name for the array (as opposed to `rs` as you suggested; making the overall change smaller, I think): https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/arch/mips/include/asm/ptrace.h#L35-L40
3. Commit 0abcf89ace15e28f697ed103c7d52cfa4e810977 crashes, though. Introduces SQ to store registers. Doesn’t introduce a corresponding LQ, as it shouldn’t be necessary yet. I’m guessing it might be an issue with this piece: https://github.com/frno7/linux/blob/0abcf89ace15e28f697ed103c7d52cfa4e810977/arch/mips/include/asm/stackframe.h#L29-L36

I also found https://github.com/frno7/linux/blob/0abcf89ace15e28f697ed103c7d52cfa4e810977/arch/mips/include/asm/processor.h#L244-L248 that seems to need careful deliberation? :-)

**Comment by rickgaiser** [comment]
LQ/SQ need 16 byte alignment, they silently load/store an aligned addr if it's not.

**Comment by frno7** [comment]
Crap. Great find, @rickgaiser. :-) It was `__aligned(8)` and now its `__aligned(16)`. Still bugs, though. Hmm.

**Comment by frno7** [comment]
Code works when I replace LQ/SQ with LD/SD, which is 64-bit access, in a kernel that is otherwise a 32-bit and would normally use LW/SW. I’m not sure whether GAS plays tricks and replaces SD with two SWs, though, but it shouldn’t matter in this test.

Also doing SQ followed by SD works, where whatever misalignment SQ had is overwritten by SD.

So the case for 16-byte misalignment strengthens considerably. Some part of the kernel allocating `struct pt_regs` may of course ignore the request for `__aligned(16)`. C language typing is unfortunately rather weak, as alignment requirements and other bits are easily lost.

**Comment by frno7** [comment]
https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/arch/mips/include/asm/ptrace.h#L177-L181 is an interesting piece. It looks like it’s computing the bottom of the stack (which normally grows downward), then skips 32 bytes followed by the size of `struct pt_regs` itself.

A variant of the same computation is found in https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/arch/mips/include/asm/compat.h#L121-L127

`THREAD_SIZE` is apparently some multiple of `PAGE_SIZE`: https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/arch/mips/include/asm/thread_info.h#L100

Then there are things like https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/arch/mips/power/cpu.c#L14 where the compiler should align `struct pt_regs` properly.

**Comment by frno7** [comment]
The crash happens very early during booting, so it could also be some kind of provisional allocation of `struct pt_regs` made somewhere in assembly. Found https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/arch/mips/kernel/head.S#L144-L150 with `PT_SIZE` defined as `sizeof(struct pt_regs)`. Now `union thread_union` containing the initial stack as in https://github.com/frno7/linux/blob/431277c1e1591b9fd88587e6ee8906f71784f7b5/include/linux/sched.h#L1633-L1641 might be 16-byte misaligned. Sadly, `__aligned(16)` to it still won’t boot.

**Comment by frno7** [comment]
Found and fixed the `struct pt_regs` 16-byte alignment problem on the stack. :-) https://github.com/frno7/linux/blob/1434675bce4bea73bcdaef6bf360815a1b80e566/arch/mips/include/asm/stackframe.h#L254-L259

**Comment by frno7** [comment]
Restoring multimedia registers with corresponding LQs also seem to work properly. I believe the additional R5900-specific shift amount (SA) register needs to be saved and restored, along with HI1 and LO1. @rickgaiser, are you aware of more registers needing handling? I believe VPU0 is turned off, so its registers shouldn’t need anything at the moment (issue #14).

**Comment by rickgaiser** [comment]
Congratulations on making it work! The FPU accumulator is only extra register I can think of.

VU0 has many registers.

**Comment by frno7** [comment]
@rickgaiser, handling the FPU accumulator according to https://github.com/frno7/linux/issues/3#issuecomment-3831401770 in issue #3.

> VU0 has many registers.

Will be exciting in issue #14, undoubtedly. :-)

Notes on handling SA, HI1 and LO1. Review _EE Core Instruction Set Manual_.

- Save SA with `MFSA rd` (64 bits) (p. 148).
- Save HI1 with `MFHI1 rd` (64 bits) (p. 146).
- Save LO1 with  `MFLO1 rd` (64 bits) (p. 147).
- Restore SA with `MTSA rs` (64 bits) (p. 151).
- Restore HI1 with `MTHI1 rs` (64 bits) (p. 149).
- Restore LO1 with `MTLO1 rs` (64 bits) (p. 150).

**Comment by frno7** [comment]
I made a small test program, using `fork`, to print the MM8, HI1, LO1 and SA registers with different values in the two processes:

```c
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

static void wr_mm8(uint64_t w)
{
	uint64_t __attribute__((aligned(16))) v[2] = { 0, w };

	__asm__ __volatile__ (
		"lq $8, %0"
		: /* No output */
		: "m" (v)
		: "$8");
}

static uint64_t rd_mm8()
{
	uint64_t __attribute__((aligned(16))) v[2] = { 0, 0 };

	__asm__ __volatile__ (
		"sq $8, %0"
		: "=m" (v)
		: /* No input */
		: "memory");

	return v[1];
}

#define REG(type, reg, wr, rd)						\
	static void wr_##reg(uint64_t w)				\
		{							\
		__asm__ __volatile__ (					\
			#wr " %0"					\
			: /* No output */				\
			: "r" (w)					\
			: /* Clobbers register */);			\
	}								\
									\
	static uint64_t rd_##reg()					\
	{								\
		uint64_t r;						\
									\
		__asm__ __volatile__ (					\
			#rd " %0"					\
			: "=r" (r)					\
			: /* No input */				\
			: /* No clobber */);				\
									\
		return r;						\
	}

REG(uint64_t, hi1, mthi1, mfhi1)
REG(uint64_t, lo1, mtlo1, mflo1)
REG(uint64_t, sa,  mtsa,  mfsa)

int main(int argc, char *argv[])
{
	const pid_t pid = fork();

	if (pid == -1) {
		perror("fork");
		exit(EXIT_FAILURE);
	}

	const int b = pid ? 1000 : 2000;

	wr_mm8(b + 1);
	wr_hi1(b + 2);
	wr_lo1(b + 3);
	wr_sa(b + 4);

	for (int n = 0; n < 5; n++) {
		printf("%d"
		       " mm8 %" PRId64
		       " hi1 %" PRId64
		       " lo1 %" PRId64
		       " sa %" PRId64 "\n",
			b,
			rd_mm8(),
			rd_hi1(),
			rd_lo1(),
			rd_sa());
		fflush(stdout);

		sleep(1);
	}

	return EXIT_SUCCESS;
}
```

The result is (with 1000 for the parent process, and 2000 for the child process):

```
1000 mm8 1001 hi1 1002 lo1 1003 sa 12
2000 mm8 2001 hi1 2002 lo1 2003 sa 4
1000 mm8 1001 hi1 1002 lo1 1003 sa 12
2000 mm8 2001 hi1 2002 lo1 2003 sa 4
1000 mm8 1001 hi1 1002 lo1 1003 sa 12
2000 mm8 2001 hi1 2002 lo1 2003 sa 4
1000 mm8 1001 hi1 1002 lo1 1003 sa 12
2000 mm8 2001 hi1 2002 lo1 2003 sa 4
1000 mm8 1001 hi1 1002 lo1 1003 sa 12
2000 mm8 2001 hi1 2002 lo1 2003 sa 4
```

SA is special and cannot be assigned an arbitrary number in the manner done in the test program. However, for the purpose of this test the important thing is that the two processes have different alternating values, as expected.

Linux 5.4 now supports MMI as well as HI1 and LO1 registers with the o32 ABI. :-)

Will close this issue as completed once code is committed to the `ps2-main` branch.

**Comment by frno7** [comment]
For reference, Linux kernels _without_ MMI and HI1, LO1 support print something like this:
```
1000 mm8 1001 hi1 1002 lo1 1003 sa 12
2000 mm8 2001 hi1 2002 lo1 2003 sa 4
1000 mm8 0 hi1 2002 lo1 2003 sa 0
2000 mm8 0 hi1 2002 lo1 2003 sa 0
1000 mm8 0 hi1 2002 lo1 2003 sa 0
2000 mm8 0 hi1 2002 lo1 2003 sa 0
1000 mm8 0 hi1 2002 lo1 2003 sa 0
2000 mm8 0 hi1 2002 lo1 2003 sa 0
1000 mm8 0 hi1 2002 lo1 2003 sa 0
2000 mm8 0 hi1 2002 lo1 2003 sa 0
```
One can observe several problems: MM8 and SA are zeroed when switching between processes, and HI1 and LO1 remain unchanged by the last process writing to them. All fixed now.

Merged to `ps2-main` branch.

**Comment by frno7** [comment]
Commit 77174dd4c8958a91aa5b76fb064a60eb68119e70 reverts save and restore multimedia registers. Approach taken doesn’t work for two reasons:

- SQ and LQ require 16-byte alignment, which is difficult to guarantee, for example in a general stack setting with a 4-byte natural alignment in a 32-bit kernel. Perhaps solvable with a major rewrite of MIPS exception mechanisms? The alternative approach is to save and restore multimedia registers in a separate array, in addition to the normal one.
- The kernel sometimes modifies the register array. These 32-bit writes must be sign-extended to 64 bits, in a 32-bit kernel, if and only if a register was written. LQ doesn’t sign-extend. It’s not enough to compare registers with original values, because the register must be sign-extended even if it happens to have the same 32-bit value as the original: the actual act of writing the register array somewhere in the kernel must be sign-extended. The kernel writes registers both in C and in assembly (using LONG_S).

Perhaps another kind of approach with custom handling of all exception methods would work better. There are several cases to consider, with various register save/restore characteristics, for example:

- interrupts (timers, I/O, etc.),
- traps for software emulated instructions,
- task switches,
- page faults,
- etc.

Commits:

- 587215439ba6 ("MIPS: R5900: Restore multimedia registers")
- eb95b1a3c2cb ("MIPS: R5900: Store multimedia registers, not yet restored")
- 0fa70646d618 ("MIPS: R5900: Introduce multimedia registers, not yet used")
- 508c006bf24e ("MIPS: R5900: Align pt_regs on stack with 16-byte boundary")
- 1e44905e434d ("MIPS: R5900: Abstract pt_regs main array access with gprs macro")
- 40e49edd0352 ("Revert "MIPS: R5900: Reset bits 127..64 of GPRs in RESTORE_SOME"")

**Comment by maciej-w-rozycki** [comment]
@frno7 Stack alignment is mandated to be 8 bytes for o32 and 16 bytes for n32/n64, so realignment would only be required with o32 kernel.  It's been done before and I suppose we'll need it anyway for the userland so that `SQ`/`LQ` can be used for frame accesses, so a compiler option might be due (see `-mstackrealign` with x86).

The need to sign-extend low 32-bits with o32 could be a showstopper, though perhaps could be sorted by making the kernel use `SD` for context modifications, by declaring the integer part of the GPRs `long long` in the context and updating assembly accessors accordingly.  Still I think this is the most promising way to handle the MMI context.

**Comment by frno7** [comment]
> [@frno7](https://github.com/frno7) Stack alignment is mandated to be 8 bytes for o32 and 16 bytes for n32/n64, so realignment would only be required with o32 kernel. It's been done before and I suppose we'll need it anyway for the userland so that `SQ`/`LQ` can be used for frame accesses, so a compiler option might be due (see `-mstackrealign` with x86).

@maciej-w-rozycki, I’m hoping we can dodge modifying the compiler with a new option for 16-byte alignment, by applying a method similar to the `__aligned__` directive in C:
```c
void g(void *p);

void f()
{
	char __attribute__((__aligned__(16))) c;

	g(&c);
}
```
Looking the the generated assembly for the C code above, GCC introduces a pair of SRL and SLL instructions, specifically for `__aligned__`, similar to what I attempted in
https://github.com/frno7/linux/blob/508c006bf24ece3e150f0f7fd165fd65b5394ddc/arch/mips/include/asm/stackframe.h#L250-L255

It should be enough to apply this method of alignment in places where the `struct pt_regs` is allocated on the stack in assembly code (the compiler automatically aligns C variables, as alignments are attached to their types). Grepping for `PT_SIZE` reveals a dozen or so assembly candidates in the kernel. Naturally, there are complications in the kernel. For example, its use of registers is often severely limited before some of them can be saved in said `struct pt_regs`. It also needs to save the user stack, while simultaneously switching to its own kernel stack (`irq_stack`).

It would be nice, I think, if a common set of macros could be made. For example something like `ALIGN_STACK(alignment_of(struct pt_regs))` that could be added to the relevant assembly routines as needed. Of course, there are further complications. For example, `PT_SIZE` can no longer be used directly as `framsize` for the `.frame` directive (p. 7-3 in _MIPS Assembly Language Programmer’s Guide_) due to the need for additional alignment padding, as seen here via the `NESTED` macro:

https://github.com/frno7/linux/blob/508c006bf24ece3e150f0f7fd165fd65b5394ddc/arch/mips/kernel/genex.S#L175

Even so, I think one can hope that a careful application of stack alignment macros in a handful or so exception handlers would be a fairly limited change, don’t you think?

Even better, it would be very nice if most, or even all, MIPS assembly exception routines could be replaced with plain C code using the [`interrupt` function attribute](https://gcc.gnu.org/onlinedocs/gcc/MIPS-Function-Attributes.html#index-interrupt-function-attribute_002c-MIPS), wouldn’t it?

> The need to sign-extend low 32-bits with o32 could be a showstopper, though perhaps could be sorted by making the kernel use `SD` for context modifications, by declaring the integer part of the GPRs `long long` in the context and updating assembly accessors accordingly. Still I think this is the most promising way to handle the MMI context.

Yes, I’ve come to the conclusion that writing to the register array must be done via SD in a 32-bit R5900 kernel. In C, reading and writing the register array seem to require separate macros, with appropriate casts for writing it (for example the combination `(int64_t)(int32_t)`). In assembly, SD should probably be preceded by SLL (as in `sll \reg, \reg, 0`) to force 32-bit sign-extensions in all cases. One might want to rename `LONG_L` and `LONG_S` to `REG_L` and `REG_S`, respectively, for clarity.

**Comment by maciej-w-rozycki** [comment]
> Even better, it would be very nice if most, or even all, MIPS assembly exception routines could be replaced with plain C code using the [`interrupt` function attribute](https://gcc.gnu.org/onlinedocs/gcc/MIPS-Function-Attributes.html#index-interrupt-function-attribute_002c-MIPS), wouldn’t it?

I do believe this attribute implies/requires MIPSr2 ISA for code generation.

> Yes, I’ve come to the conclusion that writing to the register array must be done via SD in a 32-bit R5900 kernel. In C, reading and writing the register array seem to require separate macros, with appropriate casts for writing it (for example the combination `(int64_t)(int32_t)`). In assembly, SD should probably be preceded by SLL (as in `sll \reg, \reg, 0`) to force 32-bit sign-extensions in all cases. One might want to rename `LONG_L` and `LONG_S` to `REG_L` and `REG_S`, respectively, for clarity.

In assembly there should be no need for explicit sign-extension (with `SLL` or otherwise), because by definition of the o32 ABI all the registers will have been sign-extended already (except for the link register quirk with jump-and-link and branch-and-link operations) by virtue of using 32-bit operations only.

In C indeed we have a problem due to insisting on using `unsigned long` throughout our codebase rather than `signed long` data type to hold raw register values even though everyone knows that MIPS registers are signed ever since expanding the ISA to 64-bits.  So accessors may be inevitable unless we want to turn all our code upside down.

**Comment by frno7** [comment]
> > Even better, it would be very nice if most, or even all, MIPS assembly exception routines could be replaced with plain C code using the [`interrupt` function attribute](https://gcc.gnu.org/onlinedocs/gcc/MIPS-Function-Attributes.html#index-interrupt-function-attribute_002c-MIPS), wouldn’t it?
> 
> I do believe this attribute implies/requires MIPSr2 ISA for code generation.

What a pity. Although, in its current form judging by the manual, it doesn’t seem powerful enough for general kernel use. I think it would be quite nice if it was extended with register read/write accessors, etc. :-)

> In assembly there should be no need for explicit sign-extension (with `SLL` or otherwise), because by definition of the o32 ABI all the registers will have been sign-extended already (except for the link register quirk with jump-and-link and branch-and-link operations) by virtue of using 32-bit operations only.

True, in the absence of bugs, it’s a precaution that’s unnecessary.

> In C indeed we have a problem due to insisting on using `unsigned long` throughout our codebase rather than `signed long` data type to hold raw register values even though everyone knows that MIPS registers are signed ever since expanding the ISA to 64-bits. So accessors may be inevitable unless we want to turn all our code upside down.

Agreed!
